garage/tf/models/base.py

```diff
@@ -0,0 +1,297 @@
+    be used.
+
+    ***
+    Do not call tf.global_variable_initializers() after building a model as it
+    will reassign random weights to the model.
+    The parameters inside a model will be initialized when calling build().
+    ***
+
+    Pickling is handled automatcailly. The target weights should be assigned to
+    self._default_parameters before pickling, so that the newly created model
+    can check if target weights exist or not. When unpickled, the unserialized
+    model will load the weights from self._default_parameters.
+
+    The design is illustrated as the following:
+
+         input_1                      input_2
+            |                            |
+    ============== Model (TfModel)===================
+    |       |                            |          |
+    |       |            Parameters      |          |
+    |    =============  /           \  ============ |
+    |    |  default  | /             \ | Network2 | |
+    |    | (Network) |/               \|(Network) | |
+    |    =============                 ============ |
+    |       |                            |          |
+    =================================================
+            |                            |
+            |                            |
+    (model.networks['default'].outputs)  |
+                        model.networks['Network2'].outputs
+
+
+    Examples are also available in tests/garage/tf/models/test_model.
+
+    Args:
+      name (str): Name of the model. It will also become the variable scope
+      of the model. Every model should have a unique name.
+    """
+
+    def __init__(self, name):
+        super().__init__()
+        self._name = name or type(self).__name__  # name default to class
+        self._networks = {}
+        self._default_parameters = None
+
+    def build(self, *inputs, name=None):
+        """
+        Build a Network with the given input(s).
+
+        ***
+        Do not call tf.global_variable_initializers() after building a model
+        as it will reassign random weights to the model.
+        The parameters inside a model will be initialized when calling build().
+        ***
+
+        It uses the same, fixed variable scope for all Networks, to ensure
+        parameter sharing. Different Networks must have an unique name.
+
+        Args:
+          inputs: Tensor input(s), recommended to be position arguments, e.g.
+            def build(self, state_input=None, action_input=None, name=None).
+          name(str): Name of the model, which is also the name scope of the
+            model.
+
+        Raises:
+          ValueError when a Network with the same name is already built.
+
+        Returns:
+          outputs: Output tensor of the model with the given inputs.
+
+        """
+        network_name = name or 'default'
+
+        if not self._networks:
+            _variable_scope = tf.variable_scope(self._name, reuse=False)
```

Comment:
store variable_scope rather than recreating it every time

Reply:
It will make `Model` unpicklable as `tf.variable_scope` is not picklable. But we can remove this field in `__getstate__` and construct it in `__setstate__`.